File: src/main/java/com/exampledigisphere/vitapetcare/config/security/JwtTokenUtil.java
package com.exampledigisphere.vitapetcare.config.security;

import com.exampledigisphere.vitapetcare.config.root.Info;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Collection;
import java.util.Date;

@Component
@Info(
  dev = Info.Dev.heltonOliveira,
  label = Info.Label.doc,
  date = "29/12/2025",
  description = "Utilitário para manipulação de tokens JWT"
)
public class JwtTokenUtil {

  @Value("${jwt.secret}")
  private String secret;

  private final long accessExpiration = 1000L * 60 * 60 * 24;
  private final long refreshExpiration = 1000L * 60 * 60 * 24 * 30;

  private SecretKey getSignInKey() {
    byte[] keyBytes = Decoders.BASE64.decode(secret);
    return Keys.hmacShaKeyFor(keyBytes);
  }

  public String generateAccessToken(String username, Collection<? extends GrantedAuthority> roles) {
    return Jwts.builder()
      .subject(username)
      .claim("roles", roles.stream().map(GrantedAuthority::getAuthority).toList())
      .expiration(new Date(System.currentTimeMillis() + accessExpiration))
      .signWith(getSignInKey())
      .compact();
  }

  public String generateRefreshToken(String username) {
    return Jwts.builder()
      .subject(username)
      .issuedAt(new Date())
      .expiration(new Date(System.currentTimeMillis() + refreshExpiration))
      .signWith(getSignInKey())
      .compact();
  }

  public boolean validate(String token) {
    try {
      Jwts.parser()
        .verifyWith(getSignInKey())
        .build()
        .parseSignedClaims(token);
      return true;
    } catch (Exception e) {
      return false;
    }
  }

  public String extractUsername(String token) {
    return Jwts.parser()
      .verifyWith(getSignInKey())
      .build()
      .parseSignedClaims(token)
      .getPayload()
      .getSubject();
  }
}

File: src/main/kotlin/com/exampledigisphere/vitapetcare/config/security/JwtTokenUtil.kt
package com.exampledigisphere.vitapetcare.config.security

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.core.GrantedAuthority
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey

@Component
class JwtTokenUtil {

  @Value("\${jwt.secret}")
  private lateinit var secret: String

  private val accessExpiration: Long = 1000L * 60 * 60 * 24
  private val refreshExpiration: Long = 1000L * 60 * 60 * 24 * 30

  private fun getSignInKey(): SecretKey {
    val keyBytes = Decoders.BASE64.decode(secret)
    return Keys.hmacShaKeyFor(keyBytes)
  }

  fun generateAccessToken(username: String, roles: Collection<GrantedAuthority>): String {
    return Jwts.builder()
      .subject(username)
      .claim("roles", roles.map { it.authority })
      .expiration(Date(System.currentTimeMillis() + accessExpiration))
      .signWith(getSignInKey())
      .compact()
  }

  fun generateRefreshToken(username: String): String {
    return Jwts.builder()
      .subject(username)
      .issuedAt(Date())
      .expiration(Date(System.currentTimeMillis() + refreshExpiration))
      .signWith(getSignInKey())
      .compact()
  }

  fun validate(token: String?): Boolean {
    return try {
      Jwts.parser()
        .verifyWith(getSignInKey())
        .build()
        .parseSignedClaims(token)
      true
    } catch (e: Exception) {
      false
    }
  }

  fun extractUsername(token: String): String {
    return Jwts.parser()
      .verifyWith(getSignInKey())
      .build()
      .parseSignedClaims(token)
      .payload
      .subject
  }

}
